test/adapters/adapter_test_methods.rb

```diff
@@ -0,0 +1,258 @@
+    }
+
+    try_fillup.call(1.1, 0.0, false) # Oversized fillup must be refused outright
+    try_fillup.call(0.3, 0.3, true)
+    try_fillup.call(0.3, 0.6, true)
+    try_fillup.call(0.3, 0.9, true)
+    try_fillup.call(0.3, 0.9, false) # Would take the bucket to 1.2, so must be rejected
+
+    sleep(0.2) # Leak out 0.2 tokens
+
+    try_fillup.call(0.3, 1.0, true)
+    try_fillup.call(-2, 0.0, true) # A negative fillup is permitted since it will never take the bucket above capacity
+    try_fillup.call(1.0, 1.0, true) # Filling up in one step should be permitted
+  end
+
+  def test_bucket_lifecycle_allows_conditional_fillup_after_leaking_out
+    rng = Random.new(Minitest.seed)
+    12.times do |i|
+      key = rng.hex(4)
+      capacity = 30
+      leak_rate = capacity / 0.5
+
+      _, _, did_accept = adapter.add_tokens_conditionally(key: key, capacity: capacity, leak_rate: leak_rate, n_tokens: 29.6)
+      assert did_accept, "Should have accepted the topup on iteration #{i}"
+
+      _, _, did_accept = adapter.add_tokens_conditionally(key: key, capacity: capacity, leak_rate: leak_rate, n_tokens: 1)
+      refute did_accept, "Should have refused the topup on iteration #{i}"
+
+      sleep 0.6 # Spend enough time to allow the bucket to leak out completely
+      _, _, did_accept = adapter.add_tokens_conditionally(key: key, capacity: capacity, leak_rate: leak_rate, n_tokens: 1)
+      assert did_accept, "Once the bucket has leaked out to 0 the fillup should be accepted again on iteration #{i}"
+    end
+  end
+
+  def test_set_block_sets_a_block
+    key = random_key
+    now = Time.now.utc
+    block_duration_s = 2.2
+
+    assert_nil adapter.blocked_until(key: key)
+
+    set_block_result = adapter.set_block(key: key, block_for: block_duration_s)
+    assert_kind_of Time, set_block_result
+    assert_in_delta now + block_duration_s, set_block_result, 0.1
+
+    blocked_until = adapter.blocked_until(key: key)
+    assert_in_delta blocked_until, set_block_result, 0.1
+  end
+
+  def test_set_block_does_not_set_block_in_the_past
+    key = random_key
+    assert_nil adapter.blocked_until(key: key)
+    assert_raise(ArgumentError) { adapter.set_block(key: key, block_for: -20) }
+    assert_nil adapter.blocked_until(key: key)
+  end
+
+  def test_set_block_does_not_set_block_which_would_expire_immediately
+    key = random_key
+    assert_nil adapter.blocked_until(key: key)
+    assert_raise(ArgumentError) { adapter.set_block(key: key, block_for: 0) }
+    assert_nil adapter.blocked_until(key: key)
+  end
+
+  def test_prune
+    key = random_key
+    capacity = 30
+    leak_rate = capacity / 0.5
+
+    adapter.add_tokens_conditionally(key: key, capacity: capacity, leak_rate: leak_rate, n_tokens: 29.6)
+    adapter.set_block(key: key, block_for: 0.5)
+
+    sleep 0.65
+
+    # Both the leaky bucket and the block should have expired by now, and `prune` should not raise
+    adapter.prune
```

Comment:
**nitpick:** we don't assert anything in this test. Is this is intentional? What's the point then to have it?

Reply:
There is a whole discussion about `assert_nothing_raised` here https://stackoverflow.com/questions/12499259/why-doesnt-minitestspec-have-a-wont-raise-assertion - basically, Ryan posits that if your call raises something, it means the test will fail anyway (and using "assert_nothing_raised" just inflates your green dot count and gives you fake satisfaction for no reason, which is not to be had on these premises!)

That's why I left out the assertion (I think it comes with the activesupport methods, which we do not import into the test suite just now). But I could add it if you think just "calling the thing" is not telling the reader what we are doing. Or we could stick in a comment?..

Reply:
How do you feel about  `assert adapter.prune`?

Reply:
Can do - we would need the `prune` method to return a truthy value for this.